Add optional AWS Account ID to support CloudWatch cross-account observability

jbuck opened the pull request:
<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md
-->

_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

<!--
  Make sure to link the related issue for this change
  If it requires multiple PRs and/or a PR on another repo as well, please use "Relates to" instead.
-->
Fixes #7189

<!--
  Make sure to link the related PRs for changes such as documentation & Helm charts
-->
Relates to https://github.com/kedacore/keda-docs/pull/1650 and https://github.com/kedacore/testing-infrastructure/pull/187

I still need to add tests for this change, but wanted to get this up for early feedback.

dttung2905 commented:
/run-e2e aws
**Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/19151643090)

jbuck commented:
@dttung2905 Okay, I think this is ready for re-review!

For the E2E test, I saw that AWS CloudWatch GetMetricData calls to the AWS account in which metrics are generated can have the `AccountId` specified or not, and they'll have the same result:

Showing current account ID
```
% aws sts get-caller-identity --query 'Account' 
"711387098705"
```

No AccountID specified
```
% aws cloudwatch get-metric-data --start-time 2025-11-11T18:00:00Z --end-time 2025-11-11T18:05:00Z --metric-data-queries '[{"Id":"foobar","MetricStat":{"Metric":{"Namespace":"AWS/SQS","MetricName":"ApproximateNumberOfMessagesVisible","Dimensions":[{"Name":"QueueName","Value":"relay-prod-dlq"}]},"Period":60,"Stat":"Average"}}]'                           
{
    "MetricDataResults": [
        {
            "Id": "foobar",
            "Label": "ApproximateNumberOfMessagesVisible",
            "Timestamps": [
                "2025-11-11T18:04:00+00:00",
                "2025-11-11T18:03:00+00:00",
                "2025-11-11T18:02:00+00:00",
                "2025-11-11T18:01:00+00:00",
                "2025-11-11T18:00:00+00:00"
            ],
            "Values": [
                18.0,
                18.0,
                38.0,
                33.0,
                31.0
            ],
            "StatusCode": "Complete"
        }
    ],
    "Messages": []
}
```

Account ID specified, same as SQS queue account ID
```
% aws cloudwatch get-metric-data --start-time 2025-11-11T18:00:00Z --end-time 2025-11-11T18:05:00Z --metric-data-queries '[{"Id":"foobar","MetricStat":{"Metric":{"Namespace":"AWS/SQS","MetricName":"ApproximateNumberOfMessagesVisible","Dimensions":[{"Name":"QueueName","Value":"relay-prod-dlq"}]},"Period":60,"Stat":"Average"},"AccountId":"711387098705"}]'
{
    "MetricDataResults": [
        {
            "Id": "foobar",
            "Label": "ApproximateNumberOfMessagesVisible",
            "Timestamps": [
                "2025-11-11T18:04:00+00:00",
                "2025-11-11T18:03:00+00:00",
                "2025-11-11T18:02:00+00:00",
                "2025-11-11T18:01:00+00:00",
                "2025-11-11T18:00:00+00:00"
            ],
            "Values": [
                18.0,
                18.0,
                38.0,
                33.0,
                31.0
            ],
            "StatusCode": "Complete"
        }
    ],
    "Messages": []
}
```

This is not as good as a "full" E2E test setup where there are 2 AWS accounts, but it should be sufficient to test scaler functionality.

If a E2E test is desired where there are 2 AWS accounts, I will need some guidance on changes to https://github.com/kedacore/testing-infrastructure . Do I create the AWS account in Terraform code here, or will a new variable be added for that, and I only need to setup the cross-account source/sink between the two AWS accounts?

JorTurFer commented on tests/scalers/aws/aws_cloudwatch_cross_account/aws_cloudwatch_cross_account_test.go:
```diff
@@ -0,0 +1,244 @@
+func createCloudWatchClient() *cloudwatch.Client {
+	configOptions := make([]func(*config.LoadOptions) error, 0)
+	configOptions = append(configOptions, config.WithRegion(awsRegion))
+	cfg, _ := config.LoadDefaultConfig(context.TODO(), configOptions...)
```
```suggestion
	cfg, _ := config.LoadDefaultConfig(context.Background(), configOptions...)
```

wozniakjan commented on pkg/scalers/aws_cloudwatch_scaler.go:
```diff
@@ -45,8 +45,9 @@ type awsCloudwatchMetadata struct {
-	AwsEndpoint string `keda:"name=awsEndpoint, order=triggerMetadata, optional"`
+	AwsAccountID string `keda:"name=awsAccountId, order=triggerMetadata, optional"`
+	AwsRegion    string `keda:"name=awsRegion,   order=triggerMetadata;authParams"`
+	AwsEndpoint  string `keda:"name=awsEndpoint, order=triggerMetadata, optional"`
```
minor nit regarding alignment
```suggestion
	AwsAccountID string `keda:"name=awsAccountId, order=triggerMetadata, optional"`
	AwsRegion    string `keda:"name=awsRegion,    order=triggerMetadata;authParams"`
	AwsEndpoint  string `keda:"name=awsEndpoint,  order=triggerMetadata, optional"`
```

wozniakjan reviewed (commented):
lgtm

> The problem here is that we have only 1 account and I think that getting a 2nd one will need some time (if we can get a second one), so I'm not sure about the best solution for this case :/

given the core of the change lies just in passing the account ID through input parameters
```go
input.MetricDataQueries[0].AccountId = aws.String(s.metadata.AwsAccountID)
```
I'd be ok with just a unit test instead of waiting for second AWS testing account, wdyt?

jbuck commented:
Okay, I've addressed all the comments, refactored the GetMetricDataInput creation into a separate method and added 2 unit tests, one for parsing metadata and one for checking that the account ID is assigned to GetMetricDataInput. Let me know if there's any other changes I can do!

wozniakjan commented on tests/scalers/aws/aws_cloudwatch_cross_account/aws_cloudwatch_cross_account_test.go:
```diff
@@ -0,0 +1,244 @@
+      metadata:
+        awsRegion: {{.AwsRegion}}
+				awsAccountId: {{.AwsAccountID}}
+        namespace: {{.CloudWatchMetricNamespace}}
```
ah, yeah, tests failed with
```
2025-11-21T11:48:58.5230956Z     helper.go:7***6: 
2025-11-21T11:48:58.5231770Z         	Error Trace:	/__w/keda/keda/tests/helper/helper.go:7***6
2025-11-21T11:48:58.5232709Z         	            				/__w/keda/keda/tests/helper/helper.go:745
2025-11-21T11:48:58.5233623Z         	            				/__w/keda/keda/tests/helper/helper.go:8***7
2025-11-21T11:48:58.5235034Z         	            				/__w/keda/keda/tests/scalers/aws/aws_cloudwatch_cross_account/aws_cloudwatch_cross_account_test.go:***56
2025-11-21T11:48:58.5236232Z         	Error:      	Received unexpected error:
2025-11-21T11:48:58.5237930Z         	            	error: error parsing /tmp/scaledObjectTemplate***59***97605: error converting YAML to JSON: yaml: line ***: found a tab character that violates indentation
2025-11-21T11:48:58.5239366Z         	Test:       	TestCloudWatchCrossAccountScaler
2025-11-21T11:48:58.5241025Z         	Messages:   	cannot apply file - error: error parsing /tmp/scaledObjectTemplate***59***97605: error converting YAML to JSON: yaml: line ***: found a tab character that violates indentation
```

wozniakjan commented on tests/scalers/aws/aws_cloudwatch_cross_account/aws_cloudwatch_cross_account_test.go:
```diff
@@ -0,0 +1,244 @@
+        name: keda-trigger-auth-aws-credentials
+      metadata:
+        awsRegion: {{.AwsRegion}}
+        awsAccountId: {{.AwsAccountID}}
```
```
2025-11-21T14:13:40.2178746Z     helper.go:70***: Applying template: scaledObjectTemplate
2025-11-21T14:13:40.2179048Z     helper.go:7***6: 
2025-11-21T14:13:40.2179474Z         	Error Trace:	/__w/keda/keda/tests/helper/helper.go:7***6
2025-11-21T14:13:40.2179875Z         	            				/__w/keda/keda/tests/helper/helper.go:745
2025-11-21T14:13:40.2180280Z         	            				/__w/keda/keda/tests/helper/helper.go:8***7
2025-11-21T14:13:40.2180986Z         	            				/__w/keda/keda/tests/scalers/aws/aws_cloudwatch_cross_account/aws_cloudwatch_cross_account_test.go:***56
2025-11-21T14:13:40.2181369Z         	Error:      	Received unexpected error:
2025-11-21T14:13:40.2182671Z         	            	The ScaledObject "aws-cloudwatch-cross-account-test-so" is invalid: spec.triggers[0].metadata.awsAccountId: Invalid value: "integer": spec.triggers[0].metadata.awsAccountId in body must be of type string: "integer"
2025-11-21T14:13:40.2183134Z         	Test:       	TestCloudWatchCrossAccountScaler
2025-11-21T14:13:40.2184635Z         	Messages:   	cannot apply file - The ScaledObject "aws-cloudwatch-cross-account-test-so" is invalid: spec.triggers[0].metadata.awsAccountId: Invalid value: "integer": spec.triggers[0].metadata.awsAccountId in body must be of type string: "integer"
2025-11-21T14:13:40.2185677Z     helper.go:5***5: Waiting for deployment replicas to hit target. Deployment - aws-cloudwatch-cross-account-test-deployment, Current  - 0, Target - 0
```
looks like this must be explicitly cast to string
```suggestion
        awsAccountId: "{{.AwsAccountID}}"
```

jbuck commented:
@wozniakjan thank you for the test fixes and landing this PR!
